R/formatter.r

```diff
@@ -121,6 +121,89 @@ percent_format <- function() {
 #' @rdname percent_format
 percent <- percent_format()
 
+#' Bytes formatter: convert to byte measurement and display symbol.
+#'
+#' @return a function with three parameters, \code{x}, a numeric vector that
+#'   returns a character vector, \code{symbol} the byte symbol (e.g. "\code{Kb}")
+#'   desired and the measurement \code{units} (traditional \code{binary} or
+#'   \code{si} for ISI metric units).
+#' @param x a numeric vector to format
+#' @param symbol byte symbol to use. If "\code{auto}" the symbol used will be
+#'   determined by the maximum value of \code{x}. Valid symbols are
+#'   "\code{b}", "\code{K}", "\code{Mb}", "\code{Gb}", "\code{Tb}", "\code{Pb}",
+#'   "\code{Eb}", "\code{Zb}", and "\code{Yb}", along with their upper case
+#'   equivalents and "\code{iB}" equivalents.
+#' @param units which unit base to use, "\code{binary}" (1024 base) or
+#'   "\code{si}" (1000 base) for ISI units.
+#' @references Units of Information (Wikipedia) :
+#'   \url{http://en.wikipedia.org/wiki/Units_of_information}
+#' @export
+#' @examples
+#' byte_format()(sample(3000000000, 10))
+#' bytes(sample(3000000000, 10))
+byte_format <- function(symbol="auto", units="binary") {
+  function(x) bytes(x, symbol, units)
+}
+
+#' @export
+#' @rdname byte_format
+Kb <- byte_format("Kb", "binary")
+
+#' @export
+#' @rdname byte_format
+Mb <- byte_format("Mb", "binary")
+
+#' @export
+#' @rdname byte_format
+Gb <- byte_format("Gb", "binary")
+
+#' @export
+#' @rdname byte_format
+bytes <- function(x, symbol="auto", units=c("binary", "si")) {
+
+  symbol <- match.arg(symbol, c("auto",
+        "b",  "Kb",  "Mb",  "Gb",  "Tb",  "Pb",  "Eb",  "Zb",  "Yb",
+				"B",  "KB",  "MB",  "GB",  "TB",  "PB",  "EB",  "ZB",  "YB",
+				     "KiB", "MiB", "GiB", "TiB", "PiB", "EiB", "ZiB", "YiB"))
+
+  units <- match.arg(units, c("binary", "si"))
+
+  base <- switch(units, `binary`=1024, `si`=1000)
+
+  if (symbol == "auto") {
+    symbol <-
+      if      (max(x) >= (base^5)) { "Pb" }
+      else if (max(x) >= (base^4)) { "Tb" }
+      else if (max(x) >= (base^3)) { "Gb" }
+      else if (max(x) >= (base^2)) { "Kb" }
+      else if (max(x) >= (base^1)) { "Mb" }
+      else                         {  "b" }
+  }
+
+  switch(symbol,
+	   "b" =,  "B"  = paste(x,                  "bytes"),
```

Comment:
I'd prefer to do this with character indexing, and to eliminate a lot of the duplicate code. (Maybe just look at the first letter of the unit?)